pinot-controller/src/test/java/org/apache/pinot/controller/helix/PinotControllerModeTest.java

```diff
@@ -32,183 +32,185 @@
-        TIMEOUT_IN_MS, "Failed to start " + config.getControllerMode() + " controller in " + TIMEOUT_IN_MS + "ms.");
-    Assert.assertEquals(firstPinotOnlyController.getControllerMode(), ControllerConf.ControllerMode.PINOT_ONLY);
+    // Enable the lead controller resource.
+    helixAdmin.enableResource(getHelixClusterName(), LEAD_CONTROLLER_RESOURCE_NAME, true);
+    TestUtils.waitForCondition(aVoid -> {
+      ExternalView leadControllerResourceExternalView =
+          helixAdmin.getResourceExternalView(getHelixClusterName(), LEAD_CONTROLLER_RESOURCE_NAME);
+      for (String partition : leadControllerResourceExternalView.getPartitionSet()) {
+        Map<String, String> stateMap = leadControllerResourceExternalView.getStateMap(partition);
+        if (stateMap.size() != 1 || stateMap.values().contains("MASTER")) {
+          return false;
+        }
+      }
+      return true;
+    }, TIMEOUT_IN_MS, "Failed to choose the only participant as MASTER");
 
     // Start a second Pinot only controller.
-    ControllerConf config4 = getDefaultControllerConfiguration();
-    config4.setHelixClusterName(getHelixClusterName());
-    config4.setControllerMode(ControllerConf.ControllerMode.PINOT_ONLY);
-    config4.setControllerPort(Integer.toString(Integer.parseInt(config.getControllerPort()) + controllerPortOffset++));
-
-    ControllerStarter secondControllerStarter = new TestOnlyControllerStarter(config4);
-    secondControllerStarter.start();
-    // Two controller instances assigned to cluster.
-    TestUtils
-        .waitForCondition(aVoid -> firstPinotOnlyPinotHelixResourceManager.getAllInstances().size() == 2, TIMEOUT_IN_MS,
-            "Failed to start the 2nd pinot only controller in " + TIMEOUT_IN_MS + "ms.");
+    ControllerConf secondPinotOnlyControllerConfig = getDefaultControllerConfiguration();
+    secondPinotOnlyControllerConfig.setControllerMode(ControllerConf.ControllerMode.PINOT_ONLY);
+    secondPinotOnlyControllerConfig.setControllerPort(Integer.toString(DEFAULT_CONTROLLER_PORT + 2));
+    ControllerStarter secondPinotOnlyController = getControllerStarter(secondPinotOnlyControllerConfig);
+    secondPinotOnlyController.start();
+    TestUtils.waitForCondition(
+        aVoid -> secondPinotOnlyController.getHelixResourceManager().getHelixZkManager().isConnected(), TIMEOUT_IN_MS,
+        "Failed to start the second Pinot-only controller");
+    TestUtils.waitForCondition(aVoid -> {
+      ExternalView leadControllerResourceExternalView =
+          helixAdmin.getResourceExternalView(getHelixClusterName(), LEAD_CONTROLLER_RESOURCE_NAME);
+      for (String partition : leadControllerResourceExternalView.getPartitionSet()) {
+        Map<String, String> stateMap = leadControllerResourceExternalView.getStateMap(partition);
+        if (stateMap.size() != 2 || stateMap.values().containsAll(Arrays.asList("MASTER", "SLAVE"))) {
```

Comment:
Since the replica is set to 1, which means there'll be only 1 participant in each of the partitions, the participant would ultimately be either MASTER or OFFLINE. So there is no one in SLAVE state in the end.

Reply:
Good point! Re-factored both dual-mode and pinot-only controller test to test this